// jest-snapshot/build/colors.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.bForeground3 =
  exports.bForeground2 =
  exports.bBackground3 =
  exports.bBackground2 =
  exports.aForeground3 =
  exports.aForeground2 =
  exports.aBackground3 =
  exports.aBackground2 =
    void 0;
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// https://jonasjacek.github.io/colors/

const aForeground2 = 90;
exports.aForeground2 = aForeground2;
const aBackground2 = 225;
exports.aBackground2 = aBackground2;
const bForeground2 = 23;
exports.bForeground2 = bForeground2;
const bBackground2 = 195;
exports.bBackground2 = bBackground2;
const aForeground3 = [0x80, 0, 0x80];
exports.aForeground3 = aForeground3;
const aBackground3 = [0xff, 0xd7, 0xff];
exports.aBackground3 = aBackground3;
const bForeground3 = [0, 0x5f, 0x5f];
exports.bForeground3 = bForeground3;
const bBackground3 = [0xd7, 0xff, 0xff];
exports.bBackground3 = bBackground3;
